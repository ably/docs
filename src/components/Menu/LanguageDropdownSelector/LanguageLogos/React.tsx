import React from 'react';

// The best one.
export const ReactJS = () => (
  <svg width="24" height="24" viewBox="-11.5 -10.23174 23 20.46348" fill="none" xmlns="http://www.w3.org/2000/svg">
    <circle cx="0" cy="0" r="2.05" fill="#61dafb" />
    <g stroke="#61dafb" strokeWidth="1" fill="none">
      <ellipse rx="11" ry="4.2" />
      <ellipse rx="11" ry="4.2" transform="rotate(60)" />
      <ellipse rx="11" ry="4.2" transform="rotate(120)" />
    </g>
  </svg>
);
