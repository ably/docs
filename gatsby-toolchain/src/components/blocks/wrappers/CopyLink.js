import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { ChildPropTypes } from '../../../react-utilities';
import styled from 'styled-components';
import AILink from '../../../styles/svg/ai-link';
import { secondary, tertiary } from '../../../styles/colors';

const StyledLinkCopyButton = styled.button`
  height: 100%;
  :hover {
    color: ${secondary.actionBlue};
  }
`;

const LinkHoverPopup = styled.div`
  display: inline-block;
  position: absolute;
  background-color: ${({ copySuccess }) =>
    copySuccess ? tertiary.backgroundGreen : copySuccess === null ? '#fff' : tertiary.warningRed};
  box-shadow: 0px 2px 20px 0px rgba(0, 0, 0, 0.3);
  border-radius: 4px;
  color: ${secondary.cableGrey};
  font-weight: 400;
  padding: 25px 20px;
  line-height: 100%;
  white-space: nowrap;
  word-break: initial;
  z-index: 5;
  top: 48px;
  transition: background-color 1000ms linear;

  &:after {
    content: '';
    position: absolute;
    left: 10px;
    width: 0;
    height: 0;
    bottom: 64px;
    border-left: 10px solid transparent;
    border-right: 10px solid transparent;
    border-bottom: 10px solid
      ${({ copySuccess }) =>
        copySuccess ? tertiary.backgroundGreen : copySuccess === null ? '#fff' : tertiary.warningRed};
    z-index: inherit;
    transition: border-bottom 1000ms linear;
  }
`;

const LinkCopyButton = ({ id, ...props }) => {
  const [hover, setHover] = useState(true);
  const [content, setContent] = useState('Copy section link to clipboard');
  const [copySuccess, setCopySuccess] = useState(null);

  const resetState = () => {
    setContent('Copy section link to clipboard');
    setCopySuccess(null);
  };

  useEffect(() => {
    const resetStateTimeout = setTimeout(resetState, 2000);
    return () => clearTimeout(resetStateTimeout);
  }, [copySuccess]);
  return (
    <div className="relative">
      {hover && (
        <LinkHoverPopup id={'link-copy-tooltip'} role="tooltip" copySuccess={copySuccess}>
          {content}
        </LinkHoverPopup>
      )}
      <StyledLinkCopyButton
        {...props}
        aria-describedby={'link-hover-tooltip'}
        id={id}
        onMouseEnter={() => setHover(true)}
        onMouseLeave={() => setHover(false)}
        onClick={() => {
          const linkToCopy = `${window.location.href}#${id}`;
          navigator.clipboard.writeText(linkToCopy).then(
            () => {
              setContent('Section link copied ✓');
              setCopySuccess(true);
              setTimeout(resetState, 2000);
            },
            (err) => {
              setContent(`Failed to copy section link ×\n\n${err}`);
              setCopySuccess(false);
              setTimeout(resetState, 2000);
            },
          );
        }}
      >
        <AILink />
      </StyledLinkCopyButton>
    </div>
  );
};

LinkCopyButton.propTypes = {
  id: PropTypes.string,
};

const CopyLink = ({ attribs, children }) => {
  const [hover, setHover] = useState(false);
  if (!attribs || !attribs.id) {
    return <>{children}</>;
  }
  return (
    <div className="flex items-center" onMouseEnter={() => setHover(true)} onMouseLeave={() => setHover(false)}>
      {hover && <LinkCopyButton title="Copy section link" id={attribs.id} />}
      {children}
    </div>
  );
};

CopyLink.propTypes = {
  attribs: PropTypes.object,
  children: ChildPropTypes,
};

export { LinkHoverPopup };
export default CopyLink;
