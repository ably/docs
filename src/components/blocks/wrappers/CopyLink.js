import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { ChildPropTypes } from '../../../react-utilities';
import styled from 'styled-components';
import AILink from '../../../styles/svg/ai-link';
import { secondary, tertiary } from '../../../styles/colors';

const StyledLinkCopyButton = styled.button`
  height: 100%;
  :hover {
    color: ${secondary.black};
  }
  :focus {
    outline: none;
  }
`;

const LinkHoverPopup = styled.div`
  display: inline-block;
  position: absolute;
  background-color: ${({ copySuccess }) =>
    copySuccess ? tertiary.backgroundGreen : copySuccess === null ? '#fff' : tertiary.warningRed};
  box-shadow: 0px 2px 20px 0px rgba(0, 0, 0, 0.3);
  border-radius: 4px;
  color: ${secondary.cableGrey};
  font-weight: 400;
  padding: 25px 20px;
  line-height: 100%;
  white-space: nowrap;
  word-break: initial;
  z-index: 5;
  top: 48px;
  transition: background-color 1000ms linear;

  &:after {
    content: '';
    position: absolute;
    left: 4px;
    width: 0;
    height: 0;
    bottom: 64px;
    border-left: 10px solid transparent;
    border-right: 10px solid transparent;
    border-bottom: 10px solid
      ${({ copySuccess }) =>
        copySuccess ? tertiary.backgroundGreen : copySuccess === null ? '#fff' : tertiary.warningRed};
    z-index: inherit;
    transition: border-bottom 1000ms linear;
  }
`;

const LinkCopyButton = ({ id, parentHovered, ...props }) => {
  const [hover, setHover] = useState(false);
  const [content, setContent] = useState('Copy section link to clipboard');
  const [copySuccess, setCopySuccess] = useState(null);

  const resetState = () => {
    setContent('Copy section link to clipboard');
    setCopySuccess(null);
  };

  const copyLink = () => {
    const linkToCopy = `${window.location.href.replace(/#.*$/, '')}#${id}`;
    navigator.clipboard.writeText(linkToCopy).then(
      () => {
        setContent('Section link copied ✓');
        setCopySuccess(true);
      },
      (err) => {
        setContent(`Failed to copy section link ×\n\n${err}`);
        setCopySuccess(false);
      },
    );
  };

  useEffect(() => {
    const resetStateTimeout = setTimeout(resetState, 2000);
    return () => clearTimeout(resetStateTimeout);
  }, [copySuccess]);
  return (
    <div className="relative -mt-16" onKeyPress={(event) => event.key === 'Enter' && copyLink()}>
      {(hover || parentHovered) && (
        <LinkHoverPopup id={'link-copy-tooltip'} role="tooltip" copySuccess={copySuccess}>
          {content}
        </LinkHoverPopup>
      )}
      <StyledLinkCopyButton
        {...props}
        aria-describedby={'link-hover-tooltip'}
        id={id}
        tabIndex="0"
        onMouseEnter={() => setHover(true)}
        onMouseLeave={() => setHover(false)}
        onFocus={() => setHover(true)}
        onBlur={() => setHover(false)}
        onClick={copyLink}
      >
        <AILink />
      </StyledLinkCopyButton>
    </div>
  );
};

LinkCopyButton.propTypes = {
  id: PropTypes.string,
  parentHovered: PropTypes.bool,
};

const CopyLink = ({ attribs, children }) => {
  const [hover, setHover] = useState(false);
  if (!attribs || !attribs.id) {
    return <>{children}</>;
  }
  return (
    <div className={`flex items-center`} onMouseEnter={() => setHover(true)} onMouseLeave={() => setHover(false)}>
      {children}
      <LinkCopyButton id={attribs.id} parentHovered={hover} />
    </div>
  );
};

CopyLink.propTypes = {
  attribs: PropTypes.object,
  children: ChildPropTypes,
};

export { LinkHoverPopup };
export default CopyLink;
