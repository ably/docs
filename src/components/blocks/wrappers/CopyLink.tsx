import React, { ReactNode, useState } from 'react';
import cn from 'classnames';
import Icon from '@ably/ui/core/Icon';

import { container, button, isHidden, isVisible, notification, buttonContainer } from './CopyLink.module.css';

const CopyLink = ({
  attribs,
  marginBottom = '',
  children,
}: {
  attribs?: { id?: string };
  marginBottom?: string;
  children: ReactNode;
}) => {
  const [notificationIsVisible, setNotificationIsVisible] = useState(false);

  if (!attribs || !attribs.id) {
    return <>{children}</>;
  }

  const handleCopyLink = () => {
    const linkToCopy = `${window.location.href.replace(/#.*$/, '')}#${attribs.id}`;
    navigator.clipboard.writeText(linkToCopy);
    setNotificationIsVisible(true);
    setTimeout(() => {
      setNotificationIsVisible(false);
    }, 1000);
  };

  return (
    <div className={cn(container, marginBottom)}>
      {children}
      <div className={buttonContainer}>
        <div
          className={cn(notification, {
            [isVisible]: notificationIsVisible,
          })}
        >
          Copied!
        </div>
        <button
          onClick={handleCopyLink}
          className={cn(button, {
            [isHidden]: notificationIsVisible,
          })}
        >
          <Icon name="icon-gui-link" size="1rem" />
        </button>
      </div>
    </div>
  );
};

export default CopyLink;
