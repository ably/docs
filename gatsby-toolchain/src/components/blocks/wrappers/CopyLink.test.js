import React from 'react';
import TestRenderer from 'react-test-renderer';
import CopyLink from './CopyLink';

const rootElement = (
  <CopyLink>
    <h1>Inner content</h1>
  </CopyLink>
);

const copyLinkRenderer = TestRenderer.create(rootElement);

describe('LocalLanguageAlternatives displays statically as expected', () => {
  it('Snapshot of LocalLanguageAlternatives is the same', () => {
    expect(copyLinkRenderer.toJSON()).toMatchSnapshot();
  });
});
