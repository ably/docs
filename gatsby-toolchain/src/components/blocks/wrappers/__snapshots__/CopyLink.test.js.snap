// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocalLanguageAlternatives displays statically as expected Snapshot of LocalLanguageAlternatives is the same 1`] = `
<h1>
  Inner content
</h1>
`;
